refactor(api): move last tip picked up tracking from context to engine

jbleon95 opened the pull request:
# Overview

This PR refactors the way the Python API tracks where a tip has been picked up from. Previously we were relying on the `InstrumentContext._last_tip_picked_up_from` to keep track of which tip rack well the current tip was picked up from, mostly for the purposes of storing it for `return_tip`. The issue with this approach is that anything that did not go through the top level instrument context, whether it is via the cores or directly in protocol engine, would lead to this not being updated and lead to either incorrect behavior for tip returns or an error.

This became more acute with the introduction of liquid classes and allowing them to return tips and keep them at the end, as this variable wasn't getting updated and we had to manually code around that.

Now the tracking of the last tip rack well has been moved entirely to the engine, where it will store it as a combination of `labware_id` and `well_name`. This is then reconstructed into a `LabwareCore` and `WellCore` by the `InstrumentCore`, which is further reconstructed into a `Well` object by the top level `InstrumentContext`. This allowed a lot of the liquid class scaffolding around this issue to be removed and those methods to be simplified.

This change also required adding this functionality to the legacy core. With no engine, in those older cores we are now simply storing the legacy labware and well cores when a pick up tip or drop tip is called, allowing this new method to work via both intefaces.

Because it was discovered a lot of older protocols rely on `_last_tip_picked_up_from`, a property of the same name was added that uses the new interface. In addition, a version gated `current_tip_source_well()` function was added to provide a version-protected method that should now be used instead.

## Test Plan and Hands on Testing

Snapshots and integration tests should cover a lot of the regression testing, but the following protocols will be tested:

```
requirements = {
	"robotType": "Flex",
	"apiLevel": "2.25"
}


metadata = {
    "protocolName":'Return tip new',
    'author':'Jeremy'
}

def run(protocol_context):
	tiprack = protocol_context.load_labware("opentrons_flex_96_tiprack_200ul", "C2")
	trash = protocol_context.load_trash_bin('A3')
	pipette_1k = protocol_context.load_instrument("flex_1channel_1000", "right", tip_racks=[tiprack])
	nest_plate = protocol_context.load_labware("nest_96_wellplate_2ml_deep", "D1")
	arma_plate = protocol_context.load_labware("armadillo_96_wellplate_200ul_pcr_full_skirt", "D3")

	pipette_1k.pick_up_tip()

	pipette_1k.return_tip()

	water_class = protocol_context.get_liquid_class("water")

	pipette_1k.transfer_with_liquid_class(
		liquid_class=water_class,
		volume=200,
		source=nest_plate.wells()[:4],
		dest=arma_plate.wells()[:4],
		new_tip="always",
		trash_location=trash,
		return_tip=True,
	)

	pipette_1k.transfer_with_liquid_class(
		liquid_class=water_class,
		volume=200,
		source=nest_plate.wells()[0],
		dest=arma_plate.wells()[0],
		new_tip="once",
		trash_location=trash,
		keep_last_tip=True
	)


	pipette_1k.return_tip()
```
```
requirements = {
    "robotType": "OT-2",
    "apiLevel": "2.13"
}

metadata = {
    "protocolName":'Return tip OT-2 2.13',
    'author':'Jeremy'
}

def run(protocol_context):
    p300rack = protocol_context.load_labware('opentrons_96_tiprack_300ul', 1)
    p300 = protocol_context.load_instrument('p300_single_gen2', 'left', tip_racks=[p300rack])

    p300.pick_up_tip()

    p300.return_tip()
```

## Changelog

- Refactored `_last_tip_picked_up_from` to use engine and cores to keep track of tip, not the instrument context

## Review requests

Does the way we reconstruct the `Well` object make sense?

## Risk assessment

Medium/high, this refactor touches a lot of long existing methods and is not gated by version. There's no new logic added but this affects not only Flex run code but OT-2 older API versions as well.

jbleon95 committed:
add well picked up from to pipette tip state update

jbleon95 committed:
added storing it to engine state and retrieving from cores

jbleon95 committed:
get rid of last_tip_picked_up and use engine state

jbleon95 committed:
clean up liquid class code now

jbleon95 committed:
add old internal variable as a property for backwards compat

SyntaxColoring commented on api/src/opentrons/protocol_engine/state/pipettes.py:
```diff
@@ -127,6 +127,9 @@ class PipetteState:
     has_clean_tips_by_id: Dict[str, bool]
+    last_tiprack_well_by_id: Dict[
+        str, Optional[Tuple[str, str]]
+    ]  # TODO make this another type?
```
Yeah, a `NamedTuple` or dataclass or something would be nice.

Also, nit: in general, can we spell it like `tip_rack` (two words) instead of `tiprack` (one word)?

SyntaxColoring commented on api/src/opentrons/protocol_engine/state/pipettes.py:
```diff
@@ -180,14 +184,21 @@ def _set_load_pipette(self, state_update: update_types.StateUpdate) -> None:
             if state_update.pipette_tip_state.tip_geometry:
                 attached_tip = state_update.pipette_tip_state.tip_geometry
+                last_well = state_update.pipette_tip_state.well_picked_up_from
+                assert last_well is not None
```
~~Can we enforce this with type checking in `state_update.py`?~~

Never mind, I think I see now why we didn't do this with type checking in `state_update.py`. We're using same update for dropping a tip as for picking up a tip, so this needs to be `None`-able for the drop-tip case.

Actually, with that in mind, does it make sense to remove this `assert` entirely? Always set the state to whatever it is in the update, even if it's `None`?

SyntaxColoring commented on api/tests/opentrons/protocol_engine/state/test_pipette_store_old.py:
If you're up for it, it would be good to test this new functionality in the combined `test_pipette_state.py` file (testing `PipetteStore` and `PipetteView` as a single unit) as opposed to `test_pipette_store_old.py`+`test_pipette_view_old.py`.

SyntaxColoring reviewed (commented):
Yep, all looks reasonable so far, thanks! No real notes.

jbleon95 committed:
tiprack to tip rack

jbleon95 committed:
move Well to types and rename to LabwareWellId

jbleon95 committed:
have pipette states use LabwareWellId type

sanni-t commented on api/src/opentrons/protocol_api/core/engine/instrument.py:
```diff
@@ -1049,6 +1049,23 @@ def get_nozzle_configuration(self) -> NozzleConfigurationType:
+    def get_last_well_tip_picked_up_from(
+        self,
+    ) -> Optional[Tuple[LabwareCore, WellCore]]:
+        tip_rack_id = (
```
Nit: this isn't really tiprack id, is it? Maybe call it `last_tip_pickup_info` or something?

sanni-t commented on api/src/opentrons/protocol_api/core/engine/instrument.py:
Will need to update the return signature for `_pick_up_tip_for_liquid_class` as well

sanni-t commented on api/src/opentrons/protocol_api/core/legacy/legacy_instrument_core.py:
```diff
@@ -252,6 +255,11 @@ def pick_up_tip(
+        # a well or labware
+        parent_labware, _ = location.labware.get_parent_labware_and_well()
+        if parent_labware is not None:
+            self._last_tip_rack_and_well = parent_labware._core, well_core  # type: ignore[assignment]
```
Does this need to account for the above versioned tip tracker behavior?

sanni-t reviewed (commented):
Looks good so far. Some questions and comments. Will finish reviewing the rest shortly.

sfoster1 commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -3140,6 +3105,26 @@ def _handle_dispense_target(
 
+    @property
+    def _last_tip_picked_up_from(self) -> Optional[labware.Well]:
+        """This exists to prevent old protocols that relied on this old internal variable from failing.
```
let's update this docstring a little:

- Explicitly mention that this is deprecated and ought not be used
- Explicitly point to the public method or property that should be used instead
   - Add such a public method or property if it does not exist
- But also add documentation for what it does, since it is after all going to be kept around

sfoster1 reviewed (approved):
Looks good and I'm glad we're getting to it. Let's update the docstring in the instrument context private method though please

jbleon95 committed:
fix function return type

jbleon95 committed:
taking a stab at doc string changes

jbleon95 committed:
use tiprack core thats already there

jbleon95 committed:
add new style state unit test

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -3140,6 +3105,31 @@ def _handle_dispense_target(
+        """
+        return self.get_last_tip_rack_well_used()
+
+    def get_last_tip_rack_well_used(self) -> Optional[labware.Well]:
```
Since this appears in the API ref, it should get a `@requires_version(2, 26)` decorator. Or if you don't want it in the ref, put `:meta private:` in the docstring.

sanni-t commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -3140,6 +3105,31 @@ def _handle_dispense_target(
+        """
+        return self.get_last_tip_rack_well_used()
+
+    def get_last_tip_rack_well_used(self) -> Optional[labware.Well]:
```
I kinda prefer the old name (without the underscore, of course) as it clearly said that's where the last tip was picked up from. Maybe we can just make this `last_tip_picked_up_from()`? Not a strong opinion though. Will leave it up to you.

ecormany replied:
I also thought this was a bit long. My first intuition was something like `current_tip_origin` but the word "origin" can be loaded with connotation we may or may not want here.

sanni-t commented on api/tests/opentrons/protocol_api_integration/test_transfer_with_liquid_classes.py:
```diff
@@ -2303,7 +2303,6 @@ def test_transfer_with_keep_last_tip(
             ),
         ]
         assert pipette_1k.has_tip
-        assert pipette_1k._last_tip_picked_up_from is not None
```
Any reason not to just use the public `get_last_tip_rack_well_used()` in these asserts instead?

jbleon95 replied:
Nope, I wrote this before adding back in the underscore property and then the public method. Will add it back in

sanni-t reviewed (approved):
Looks great! Thanks

ddcc4 commented on api/src/opentrons/protocol_api/core/instrument.py:
```diff
@@ -310,6 +310,12 @@ def get_blow_out_flow_rate(self, rate: float = 1.0) -> float:
         ...
 
+    @abstractmethod
+    def get_last_well_tip_picked_up_from(
```
Do you want to call this something shorter like `get_tip_origin()` or `get_tip_source()`?

ddcc4 replied:
Semantically, can you explain what this value should be if we picked up the last tip from somewhere, and then dropped the last tip? Would this become `None` after we drop (or returned) the tip?

ddcc4 commented on api/src/opentrons/protocol_engine/commands/pick_up_tip.py:
```diff
@@ -155,6 +155,9 @@ async def execute(
                 .update_pipette_tip_state(
                     pipette_id=pipette_id,
                     tip_geometry=e.tip_geometry,
+                    well_picked_up_from=LabwareWellId(
```
Yeah, this name is a bit clunky, and `well_picked_up_from` makes it sound like this argument just identifies a well, and not the labware.

Would a name like `tip_source` be unambigious?

ddcc4 commented on api/src/opentrons/protocol_engine/state/pipettes.py:
```diff
@@ -127,6 +128,7 @@ class PipetteState:
     liquid_presence_detection_by_id: Dict[str, bool]
     ready_to_aspirate_by_id: Dict[str, bool]
     has_clean_tips_by_id: Dict[str, bool]
+    last_tip_rack_well_by_id: Dict[str, Optional[LabwareWellId]]
```
lol, this name is super confusing if someone comes across it in isolation.

Are you sure you don't want to call it something like `tip_source_by_id`?

ddcc4 reviewed (approved):
Thanks for doing this!

My main feedback is just on naming the variables to be more obvious (which I think Ed also mentioned as well).

sanni-t commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -3140,6 +3105,31 @@ def _handle_dispense_target(
+            return None
+        labware_core, well_core = tip_rack_cores
+        tip_rack_labware = self._core_map.get(labware_core)
+        return labware.Well(
+            parent=tip_rack_labware, core=well_core, api_version=self._api_version
+        )
```
Forgot to address your question earlier-

> Does the way we reconstruct the Well object make sense?

I don't think it's a big issue. I believe we do construct new `Well` objects in a couple of other places. It just creates a slight inconvenience of not being able to use `is` identity checks on a different instance of the same well.

Did you try using an alternative like this? Were there any issues with doing this?
```suggestion
        return tip_rack_labware.wells_by_name()[well_core.get_name()]
```

SyntaxColoring replied:
Incidentally, from memory leak investigations: It would be nice if we could always construct `Well` objects lazily instead of eagerly. They can easily get into the thousands, which I suspect it's a significant chunk of protocol memory usage. (I don't have reliable measurements handy to confirm this, though, so grain of salt.)

I vaguely remember the `is` thing being something that users problematically relied on, and we've tried to preserve it? But it just seems like such a losing battle, I'd support "breaking" it, if it helps.

jbleon95 committed:
rename stuff to use tip_origin or tip_source which is more clear

jbleon95 committed:
add public method

ddcc4 commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -3140,6 +3105,35 @@ def _handle_dispense_target(
+    @property
+    def _last_tip_picked_up_from(self) -> Optional[labware.Well]:
+        """
+        .. deprecated:: 2.26
```
```suggestion
        .. deprecated:: 2.25
```

ddcc4 commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -3140,6 +3105,35 @@ def _handle_dispense_target(
+        return self._get_last_tip_rack_well_used()
+
+    @requires_version(2, 25)
+    def last_tip_rack_well_used(self) -> Optional[labware.Well]:
```
Hey :) After a bit more back and forth, we think `current_tip_source_well()` would be the best public-facing name for this function. (https://opentrons.slack.com/archives/C06M6R72UDS/p1754668289849699)

Would you be OK with changing it?

ddcc4 replied:
```suggestion
    def current_tip_source_well(self) -> Optional[labware.Well]:
```

ddcc4 commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -1070,8 +1072,9 @@ def return_tip(self, home_after: Optional[bool] = None) -> InstrumentContext:
             _log.warning("Pipette has no tip to return")
 
-        loc = self._last_tip_picked_up_from
+        loc = self._get_last_tip_rack_well_used()
```
```suggestion
        loc = self._get_current_tip_source_well()
```

ddcc4 commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -1850,7 +1848,7 @@ def transfer_with_liquid_class(
             dest=dest,
             tip_policy=new_tip,
-            last_tip_well=self._last_tip_picked_up_from,
+            last_tip_well=self._get_last_tip_rack_well_used(),
```
```suggestion
            last_tip_well=self._get_current_tip_source_well(),
```

ddcc4 commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -1990,7 +1977,7 @@ def distribute_with_liquid_class(
             dest=dest,
             tip_policy=new_tip,
-            last_tip_well=self._last_tip_picked_up_from,
+            last_tip_well=self._get_last_tip_rack_well_used(),
```
```suggestion
            last_tip_well=self._get_current_tip_source_well(),
```

ddcc4 commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -2139,7 +2115,7 @@ def consolidate_with_liquid_class(
             dest=dest,
             tip_policy=new_tip,
-            last_tip_well=self._last_tip_picked_up_from,
+            last_tip_well=self._get_last_tip_rack_well_used(),
```
```suggestion
            last_tip_well=self._get_current_tip_source_well(),
```

ddcc4 commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -3140,6 +3105,35 @@ def _handle_dispense_target(
         if isinstance(target, validation.PointTarget):
             return target.location, None, None
 
+    def _get_last_tip_rack_well_used(self) -> Optional[labware.Well]:
```
```suggestion
    def _get_current_tip_source_well(self) -> Optional[labware.Well]:
```

ddcc4 commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -3140,6 +3105,35 @@ def _handle_dispense_target(
+    def _last_tip_picked_up_from(self) -> Optional[labware.Well]:
+        """
+        .. deprecated:: 2.26
+           Use :py:obj:`ProtocolContext.last_tip_rack_well_used` instead.
```
```suggestion
           Use :py:obj:`ProtocolContext.current_tip_source_well` instead.
```

ddcc4 commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -3140,6 +3105,35 @@ def _handle_dispense_target(
+           If the pipette has a tip on it, returns the tip rack well it was picked up from.
+           Otherwise will return ``None``.
+        """
+        return self._get_last_tip_rack_well_used()
```
```suggestion
        return self._get_current_tip_source_well()
```

ddcc4 commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -3140,6 +3105,35 @@ def _handle_dispense_target(
+
+        If there is no tip currently on the pipette, this will return ``None``.
+        """
+        return self._get_last_tip_rack_well_used()
```
```suggestion
        return self._get_current_tip_source_well()
```

ddcc4 committed:
Update api/src/opentrons/protocol_api/instrument_context.py

Change deprecation docstring to 2.25, since that's the version we'll be releasing this change.

ddcc4 committed:
Apply suggestions from code review

Rename public function `last_tip_rack_well_used()` -> `current_tip_source_well()`
